cmd/object-handlers.go

```diff
@@ -492,27 +492,49 @@ func (api objectAPIHandlers) HeadObjectHandler(w http.ResponseWriter, r *http.Re
 		return
 	}
 
-	gr, err := getObjectNInfo(ctx, bucket, object, nil, r.Header)
+	// Get request range.
+	var rs *HTTPRangeSpec
+	rangeHeader := r.Header.Get("Range")
+	if rangeHeader != "" {
+		var err error
+		if rs, err = parseRequestRangeSpec(rangeHeader); err != nil {
+			// Handle only errInvalidRange. Ignore other
+			// parse error and treat it as regular Get
+			// request like Amazon S3.
+			if err == errInvalidRange {
+				writeErrorResponseHeadersOnly(w, ErrInvalidRange)
+				return
+			}
+
+			// log the error.
+			logger.LogIf(ctx, err)
+		}
+	}
+
+	gr, err := getObjectNInfo(ctx, bucket, object, rs, r.Header)
 	if err != nil {
 		writeErrorResponseHeadersOnly(w, toAPIErrorCode(err))
 		return
 	}
 	defer gr.Close()
+
 	objInfo := gr.ObjInfo
 
-	var encrypted bool
 	if objectAPI.IsEncryptionSupported() {
-		if encrypted, err = DecryptObjectInfo(objInfo, r.Header); err != nil {
-			writeErrorResponse(w, toAPIErrorCode(err), r.URL)
+		if _, err = DecryptObjectInfo(objInfo, r.Header); err != nil {
+			writeErrorResponseHeadersOnly(w, toAPIErrorCode(err))
 			return
-		} else if encrypted {
-			s3Encrypted := crypto.S3.IsEncrypted(objInfo.UserDefined)
-			if s3Encrypted {
-				w.Header().Set(crypto.SSEHeader, crypto.SSEAlgorithmAES256)
-			} else {
-				w.Header().Set(crypto.SSECAlgorithm, r.Header.Get(crypto.SSECAlgorithm))
-				w.Header().Set(crypto.SSECKeyMD5, r.Header.Get(crypto.SSECKeyMD5))
-			}
+		}
+	}
+
+	// Set encryption response headers
+	if objectAPI.IsEncryptionSupported() {
```

Comment:
For correctness: `&& crypto.IsEncrypted(objInfo.UserDefined)`
and than:
```
switch {
case crypto.S3.IsEncrypted(objInfo.UserDefined):
case crypto.SSEC.IsEncrypted(objInfo.UserDefined)
}
```

Reply:
Shouldnt we be setting different reply headers?

Reply:
Yeah, we should. My idea was:
```
switch {
default:
     // internal error response -> encrypted but neither S3 nor SSEC
case crypto.S3.IsEncrypted(objInfo.UserDefined):
	w.Header().Set(crypto.SSEHeader, crypto.SSEAlgorithmAES256)
case crypto.SSEC.IsEncrypted(objInfo.UserDefined):
        w.Header().Set(crypto.SSECAlgorithm, r.Header.Get(crypto.SSECAlgorithm))
        w.Header().Set(crypto.SSECKeyMD5, r.Header.Get(crypto.SSECKeyMD5))
}
```
thought that was implicit. My point is:
It's not reliable to use `crypto.S3.IsEncrypted / crypto.SSEC.IsEncrypted` if you haven't checked
whether the object is encrypted at all with `crypto.IsEncrypted`. As documented `crypto.IsEncrypted` returns true if **at least one** encryption metadata entry is present.
Other `IsEncrypted` functions only check entries specific for their method (S3 / SSEC / ...)

Reply:
See: https://github.com/minio/minio/blob/master/cmd/crypto/metadata.go#L38 in contrast to e.g. 
https://github.com/minio/minio/blob/master/cmd/crypto/metadata.go#L59

In can happen (in error case) that the object is marked as encrypted but not as S3-encrypted nor SSEC-encrypted.

Reply:
Done